File: src/main/java/study/codingtest/leetcode/divide_n_conquer/LC_169.ts
const majorityElement = (nums: number[]): number => {
  const majorityElement = (nums: number[], lo: number, hi: number): number => {
    if (lo === hi) {
      return nums[lo];
    }

    const mid = lo + Math.floor((hi - lo) / 2);
    const left = majorityElement(nums, lo, mid);
    const right = majorityElement(nums, mid + 1, hi);

    let countA = 0;
    for (let i = lo; i <= hi; i++) {
      if (nums[i] === left) {
        countA++;
      }
    }
    return countA > (hi - lo + 1) / 2 ? left : right;
  };

  return majorityElement(nums, 0, nums.length - 1);
};

File: src/main/java/study/codingtest/leetcode/divide_n_conquer/LC_169.kt
fun majorityElement(nums: IntArray): Int {
  fun majorityElement(left: Int, right: Int, nums: IntArray): Int {
    if (left == right) {
      return nums[left]
    }

    val mid = left + (right - left) / 2
    val a = majorityElement(left, mid, nums)
    val b = majorityElement(mid + 1, right, nums)

    var countA = 0
    for (i in left..right) {
      if (nums[i] == a) {
        countA++
      }
    }

    return if (countA > (right - left + 1) / 2) a else b
  }

  return majorityElement(0, nums.size - 1, nums)
}